File: src/main/java/com/testerhome/appcrawler/diff/XPathUtil.java
package com.testerhome.appcrawler.diff;

import java.io.IOException;
import java.io.StringReader;
import java.util.*;

import com.testerhome.appcrawler.URIElement;
import org.w3c.dom.Document;
import org.w3c.dom.NamedNodeMap;
import org.w3c.dom.Node;
import org.w3c.dom.NodeList;
import org.xml.sax.InputSource;
import org.xml.sax.SAXException;
import javax.xml.parsers.DocumentBuilderFactory;
import javax.xml.parsers.ParserConfigurationException;
import javax.xml.xpath.*;

public class XPathUtil {
	
	public static List<String> xpathExpr= Arrays.asList("name", "label", "value", "resource-id", "content-desc", "class", "text", "index");
	
	public static List<String> xpathList = Arrays.asList("//*[contains(name(), 'Text')]", "//*[contains(name(), 'Image')]", "//*[contains(name(), 'Button')]");
	
	//用例对比
	public static Map<String,String> checkDom(String mDom, String cDom, String keyd) throws Exception  {
		
//		if(keyd.equals("com.xueqiu.android.雪球.Start(Start)_depth=2")){
//			System.out.println("");
//		}
		
		Map<String, URIElement> melements=getListFromDom(mDom);
		Map<String, URIElement> celements=getListFromDom(cDom);
//		if(keyd.equals("com.xueqiu.android.雪球.Start(Start)_depth=2")){
//			for (URIElement entry : celements.values()) {
//				if(entry.tag.equals("android.widget.TextView"))
//					System.out.println(entry.text);
//			}
//		}
		Map<String,String> keys = new HashMap<String,String>();
		HashSet<String> keySet = new HashSet<String>();
		boolean flag = true;
		
		if(melements==null && celements==null) 
			flag = false;
		if(flag&&melements==null)
		{
			keySet.addAll(celements.keySet());
			Iterator<String> it = keySet.iterator();  
			while (it.hasNext()) {
				String key = it.next();
				URIElement c=celements.get(key);
				keys.put("", c.toString());
				}

			flag = false;
		}
		if(flag&&celements==null)
		{
			keySet.addAll(melements.keySet());
			Iterator<String> it = keySet.iterator();  
			while (it.hasNext()) {
				String key = it.next();
				URIElement m=melements.get(key);
				keys.put(m.toString(),"" );
				}

			flag = false;
		}
		if(flag){
			keySet.addAll(melements.keySet());
			keySet.addAll(celements.keySet());
			Iterator<String> it = keySet.iterator();  
			while (it.hasNext()) {
				String key = it.next();
				URIElement m=melements.get(key);
				URIElement c=celements.get(key);
				if(m==null&&c==null) flag = false;
				else {
					if(m==null||c==null) {
						if (m==null) keys.put("", c.xpath());
						else keys.put(m.toString(), "");
					}
					else {
						if(!(m.name().equals(c.name())&&m.id().equals(c.id())&&m.xpath().equals(c.xpath())&&m.text().equals(c.text()))) {
							keys.put(m.toString(), c.toString());
						}
					}					
				}				
			} 
		}
		return keys;			
	}
	
	//从dom中获取一组元素
	private static Map<String, URIElement> getListFromDom(String dom) throws XPathExpressionException, SAXException, ParserConfigurationException, IOException{
		
		if (dom.isEmpty()) return null;
		
		List<URIElement> elements = new ArrayList<URIElement>();
		NodeList nodeList = null;
		StringReader sr = new StringReader(dom);
		InputSource is = new InputSource(sr);
		Document document = DocumentBuilderFactory.newInstance().newDocumentBuilder().parse(is);
		
		for (int x=0;x<xpathList.size();x++) {
			String xpath = xpathList.get(x);
			XPath xPath = XPathFactory.newInstance().newXPath();
			XPathExpression xpathExp =xPath.compile(xpath);
			if (xpath.matches("string(.*)") ||xpath.matches(".*/@[^/]*")) {
				//nodeString=(String) xpathExp.evaluate(document, XPathConstants.STRING);
				return null;   
			} 
			else nodeList=(NodeList) xpathExp.evaluate(document, XPathConstants.NODESET);
			
			if(nodeList.getLength()>0) {
				for(int i=0;i<nodeList.getLength();i++) {
					URIElement temp = new URIElement();
					Node node = (Node) nodeList.item(i);
					Map<String,String> nodeMap= new HashMap<String,String>();
					nodeMap.put("name", "");
					nodeMap.put("value", "");
					nodeMap.put("label", "");
					nodeMap.put("x", "0");
					nodeMap.put("y", "0");
					nodeMap.put("width", "0");
					nodeMap.put("height", "0");
					nodeMap.put("name()", node.getNodeName());
					nodeMap.put("innerText",node.getTextContent().trim());
					
					NamedNodeMap nodeAttributes = node.getAttributes();
					for(int j=0;j<nodeAttributes.getLength();j++) {
						Node attr = nodeAttributes.item(j);
						nodeMap.put(attr.getNodeName(), attr.getNodeValue());
					}
					
					if (nodeMap.containsKey("resource-id")) {
						nodeMap.put("name", nodeMap.get("resource-id"));
			            if (nodeMap.containsKey("text")) {
			            	nodeMap.put("value", nodeMap.get("text"));
			            }
			            if (nodeMap.containsKey("content-desc")) {
			            	nodeMap.put("label", nodeMap.get("content-desc"));
			            }
					}
					
					if (nodeMap.containsKey("href")) {
			            if (nodeMap.containsKey("id")) {
			            	nodeMap.put("name", nodeMap.get("id"));
			            }
			            if (nodeMap.containsKey("name")) {
			            	nodeMap.put("label", nodeMap.get("name"));
			            }
		            	nodeMap.put("value", nodeMap.get("innerText"));
					}
					List<Map<String, String>> attributesList=getAttributesFromNode(node);
					nodeMap.put("depth", String.valueOf(attributesList.size()));
					nodeMap.put("ancestor", getAncestorFromAttributes(attributesList));
					nodeMap.put("xpath", getXPathFromAttributes(attributesList));
					if((nodeMap.getOrDefault("visible", "") == "true" &&
				            nodeMap.getOrDefault("enabled", "") == "true" &&
				            nodeMap.getOrDefault("valid", "") == "true")) {
						nodeMap.put("valid", "true");
					}
					else nodeMap.put("valid", "false");

					if (!nodeMap.get("xpath").isEmpty() && nodeMap.get("value").toString().length()<50)
					{
						temp.tag_$eq(nodeMap.getOrDefault("name()", ""));
						temp.id_$eq(nodeMap.getOrDefault("name", ""));
						temp.name_$eq(nodeMap.getOrDefault("label", ""));
						temp.text_$eq(nodeMap.getOrDefault("value", ""));
						temp.instance_$eq(nodeMap.getOrDefault("instance", ""));
						temp.depth_$eq(nodeMap.getOrDefault("depth", ""));
						temp.xpath_$eq(nodeMap.getOrDefault("xpath", ""));
						if(nodeMap.containsKey("bounds")){
				        	String[] rect=nodeMap.get("bounds").toString().split(",");
							temp.x_$eq(Integer.parseInt(rect[0].replace("[","")));
							temp.y_$eq(Integer.parseInt(rect[1].split("\\]")[0]));
							temp.width_$eq(Integer.parseInt(rect[1].split("\\[")[1]));
							temp.height_$eq(Integer.parseInt(rect[2].replace("]","")));
				          }
						temp.ancestor_$eq(nodeMap.getOrDefault("ancestor", ""));
						temp.selected_$eq(nodeMap.getOrDefault("selected", "false"));
						temp.valid_$eq(nodeMap.getOrDefault("valid", "true"));
						elements.add(temp);
					}
				}
			}
		}
		if(!elements.isEmpty())	{
			Map<String, URIElement> elementsMap = new HashMap<String, URIElement>();
			for(int i=0;i<elements.size();i++) {
				elementsMap.put(elements.get(i).xpath(), elements.get(i));
			}
			return elementsMap;
		}
		else return null;
	}
	
	//获得一个结点的所有祖先属性列表，递归
	private static List<Map<String, String>> getAttributesFromNode(Node node){
		List<Map<String, String>> attributesList =new ArrayList<Map<String, String>>();
		attributesList=getParent(attributesList,node);
		return attributesList;
	}
	
	//获得一个结点的所有祖先属性列表，递归
	private static List<Map<String, String>> getParent(List<Map<String, String>> attributesList,Node node) {
		if (node.hasAttributes()) {
			NamedNodeMap attributes = node.getAttributes();
			Map<String, String> attributeMap = new HashMap<String,String>();
			for(int i=0;i<attributes.getLength();i++) {
				Node temp = (Node) attributes.item(i);
				attributeMap.put(temp.getNodeName(), temp.getNodeValue());
			}
			attributeMap.put("name()", node.getNodeName());
			attributeMap.put("innerText", node.getTextContent().trim());
			attributesList.add(attributeMap);
		}
		if (node.getParentNode() != null) {
			attributesList=getParent(attributesList,node.getParentNode());
	      }
		return attributesList;
	}
	
	//从属性列表中获取一个元素的Ancestor属性
	private static String getAncestorFromAttributes(List<Map<String, String>> attributesList) {
		List<String> idList=new ArrayList<String>();
		List<String> tagList=new ArrayList<String>();
		String idString = "";
		String tagString = "";
		Iterator<Map<String, String>> attributesHead = attributesList.iterator();
		while(attributesHead.hasNext()){
			String id="";
			Map<String, String> attributes = attributesHead.next();
			if(attributes.containsKey("name")) id=attributes.get("name");
			String resourceid="";
			if(attributes.containsKey("resource-id")) resourceid=attributes.get("resource-id");
			String[] idSplit= resourceid.split("/");
			id=id+idSplit[idSplit.length-1];
			if(!id.isEmpty()) idList.add(id);
			if(attributes.containsKey("name()")) tagList.add(attributes.get("name()"));
		}
		if (idList.size()>0) {
			idString=idString+idList.get(0);
			for(int i=1;i<idList.size();i++) idString=idString+"."+idList.get(i);
		}
		if (tagList.size()>0) {
			tagString=tagString+tagList.get(0);
			for(int i=1;i<idList.size();i++) tagString=tagString+"/"+tagList.get(i);
		}
		if(tagString.isEmpty()) return idString;
		else {
			if(idString.isEmpty()) return tagString;
			else return idString+"_"+tagString;
		}
	}
	
	//确定一个元素的xpath属性
	@SuppressWarnings({ "rawtypes", "unused"})
	private static String getXPathFromAttributes(List<Map<String, String>> attributes) {
		int index = attributes.size();
		List<String> xpath= new ArrayList<String>();
		Iterator<Map<String, String>> attributesHead = attributes.iterator();
		while(attributesHead.hasNext()){
			Map<String, String> attributeMap = attributesHead.next();
			index-=1;
			for(int i=0;i<xpathExpr.size();i++) {
				if(attributeMap.get(xpathExpr.get(i))==null) 
					{attributeMap.remove(xpathExpr.get(i));}
				else {
					if(attributeMap.containsKey("path"))
						attributeMap.remove("path");
				}
			}
			if (attributeMap.get("name") == attributeMap.get("label"))  
				attributeMap.remove("name");
			if (attributeMap.get("content-desc") == attributeMap.get("resource-id")) 
				attributeMap.remove("resource-id");
			String tag = "";
			if(xpathExpr.contains("name()")&&!attributeMap.get("name()").isEmpty()) 
				tag=attributeMap.get("name()");
		    else tag="*";
			Iterator<?> head = attributeMap.entrySet().iterator();
			String xpathAttributes = "";
			while(head.hasNext()){
				String temp = "";
				Map.Entry attribut = (Map.Entry) head.next();
				String key = attribut.getKey().toString();
				String value = attribut.getValue().toString();
				boolean flag = true;
				if (key.equals("name()")) {temp="";flag=false;}
				if (key.equals("innerText")) {
					if(xpathExpr.contains("innerText") && index==0 ){
						temp="contains(text()," + "'" + value.trim().substring(0, 19).replace("\"", "\\\"") + "')";
			            }else{
			            	temp="";
			            }
					flag=false;
				}
				if (key.equals("name")) {if (value.length()>50) temp="";flag=false;}
				if (key.equals("text")) {
					if (attributeMap.get("name()").contains("Button")==false && value.length()>10)
					temp="";
					flag=false;
	          }
				if(flag) {
					if(xpathExpr.contains(key)&& !value.isEmpty()) temp="@"+key+"='"+value.replace("\"", "\\\"")+ "'";
				}
				if (!temp.isEmpty()) {
					if(xpathAttributes.isEmpty()) 
						xpathAttributes=temp;
					else {
						xpathAttributes=xpathAttributes+" and "+temp;
					}
				}
			}
			if(!xpathAttributes.isEmpty()) {
				String temp ="//"+tag+"["+xpathAttributes+"]";
				xpath.add(temp);
			}				
		}
//		if(!(xpath==null)&&xpath.size()>1) xpath.remove(0);
//		boolean shortXPath=false;
//		int k=0;
//		for(int i=0;i<xpath.size();i++) {
//			if(shortXPath==false){
//				if(xpath.get(i).contains(" and ")) k++;
//					if(k>2) shortXPath=true;
//				}
//			else {
//				xpath.remove(i);
//				i--;
//			}
//		}
		if(xpath==null) return "";
		else return xpath.toString();
	}
}

File: src/main/scala/com/ceshiren/appcrawler/utils/XPathUtil.scala
package com.ceshiren.appcrawler.utils

import com.ceshiren.appcrawler.utils.Log.log
import com.fasterxml.jackson.databind.SerializationFeature
import com.fasterxml.jackson.dataformat.xml.XmlMapper
import org.w3c.dom.{Attr, Document, Node, NodeList}
import org.xml.sax.InputSource

import java.io.StringReader
import javax.xml.parsers.{DocumentBuilder, DocumentBuilderFactory}
import javax.xml.xpath.{XPath, XPathConstants, XPathFactory}
import scala.collection.mutable
import scala.collection.mutable.ListBuffer


/**
  * Created by seveniruby on 16/3/26.
  */
object XPathUtil {
  var xpathExpr = List("name", "label", "value", "resource-id", "content-desc", "class", "text", "index")
  var tagList = List("class", "resource-id", "name", "content-desc", "label")

  val builderFactory: DocumentBuilderFactory = DocumentBuilderFactory.newInstance()
  val builder: DocumentBuilder = builderFactory.newDocumentBuilder()
  val mapper = new XmlMapper

  def toDocument(raw: String): Document = {
    //todo: appium有bug, 会返回&#非法字符. 需要给appium打补丁
    //todo: 解决html的兼容问题 SAXParseException: 注释中不允许出现字符串 "--"
    val rawFormat = raw.replaceAll("[\\x00-\\x1F]", "")
      .replaceAll("&", "_")
    try {
      val document: Document = builder.parse(
        new InputSource(
          new StringReader(rawFormat)))
      document
    } catch {
      case ex: Exception => {
        log.error(rawFormat)
        scala.reflect.io.File("/tmp/1.xml").writeAll(rawFormat)
        throw ex
      }
    }
  }

  //todo: xml中有$b
  def toPrettyXML(raw: String): String = {
    val xmlNode=mapper.readTree(raw)
    mapper.enable(SerializationFeature.INDENT_OUTPUT)
    mapper.writeValueAsString(xmlNode)


/*    //done: android page source不能格式化，但是普通的xml可以, android page source在1.9后多了一些多余的空格
    val document = toDocument(raw.trim.replaceAll("> *<", "><"))
    //done: 不支持java10, use Xalan replace
    val transformerFactory = TransformerFactory.newInstance
    val transformer = transformerFactory.newTransformer
    transformer.setOutputProperty(OutputKeys.INDENT, "yes")
    transformer.setOutputProperty("{http://xml.apache.org/xslt}indent-amount", "2")
    transformer.setOutputProperty(OutputKeys.OMIT_XML_DECLARATION, "yes")
    val source = new DOMSource(document)
    val strWriter = new StringWriter
    val result = new StreamResult(strWriter)

    transformer.transform(source, result)

    return strWriter.getBuffer.toString*/

  }

  def setXPathExpr(expr: List[String]): Unit = {
    xpathExpr = expr
  }
  def setTagList(expr: List[String]): Unit = {
    tagList = expr
  }

  /**
    * 从属性中获取xpath的唯一表达式
    *
    * @param attributes
    * @return
    */
  def getXPathFromAttributes(attributes: ListBuffer[Map[String, String]], includeList: List[String]): String = {
    var index = attributes.size
    val xpath = attributes.map(attribute => {
      index -= 1
      var newAttribute = attribute
      //如果有值就不需要path了, 基本上两层xpath定位即可唯一
      includeList.foreach(key => {
        if (newAttribute.getOrElse(key, "").isEmpty) {
          newAttribute = newAttribute - key
        } else {
          newAttribute = newAttribute - "path"
        }
      })

      val tag = if (includeList.contains("name()")) {
        newAttribute.getOrElse("name()", "*")
      } else {
        "*"
      }
      val xpathAttributes = newAttribute.map(kv => {
        //todo: appium的bug. 如果控件内有换行getSource会自动去掉换行. 但是xpath表达式里面没换行会找不到元素
        //todo: 需要帮appium打补丁

        kv._1 match {
          //不同的实现name()可能会有差异，所以不考虑这个，尽量使用class属性
          case "name()" => ""
          case "class" if kv._2.contains("hierarchy")=> ""
          case "innerText" => {
            if (includeList.contains("innerText") && index == 0) {
              s"contains(text()," + "'" + kv._2.trim.take(20).replace("\"", "\\\"") + "')"
            } else {
              ""
            }

          }
          //case "index" => ""
          case "name" if kv._2.size > 50 => ""
          //todo: 优化长文本的展示
          case "text" if !newAttribute("name()").contains("Button") && kv._2.length > 10 => {
            ""
          }
          case key if includeList.contains(key) && kv._2.nonEmpty =>
            s"@${kv._1}=" + "'" + kv._2.replace("\"", "\\\"") + "'"
          case _ => ""
        }
      }).filter(x => x.nonEmpty).mkString(" and ")

      //todo: 改进xpath算法，更灵活，更唯一，更简短
      //todo: macaca的source有问题
      if (xpathAttributes.isEmpty) {
        ""
      } else {
        s"//${tag}[${xpathAttributes}]"
      }
    }
    )

    //    var shortXPath = false
    //    xpath.indices.reverse.foreach(i => {
    //      if (!shortXPath) {
    //        if (xpath(i).contains(" and ")) {
    //          shortXPath = true
    //        }
    //      } else {
    //        xpath.remove(i)
    //      }
    //    })

    xpath.mkString("")
  }

  def getAttributesFromNode(node: Node): ListBuffer[Map[String, String]] = {
    val attributesList = ListBuffer[Map[String, String]]()

    //递归获取路径,生成可定位的xpath表达式
    def getParent(node: Node): Unit = {
      if (node.hasAttributes) {
        val attributes = node.getAttributes
        var attributeMap = Map[String, String]()

        0 until attributes.getLength foreach (i => {
          val kv = attributes.item(i).asInstanceOf[Attr]
          attributeMap ++= Map(kv.getName -> kv.getValue)
        })
        attributeMap ++= Map("name()" -> node.getNodeName)
        attributeMap ++= Map("innerText" -> node.getTextContent.trim)
        attributesList += attributeMap
      }

      if (node.getParentNode != null) {
        getParent(node.getParentNode)
      }
    }

    getParent(node)
    //返回一个从root到leaf的属性列表
    return attributesList.reverse

  }


  def getAncestorFromNode(node: Node): Unit = {


  }

  def isMenuFromBrotherNode(node: Node): Boolean = {
    var b = false
    val nodeList = node.getParentNode.getChildNodes
    0 until (nodeList.getLength) foreach (i => {
      if (b == false) {
        val attributes = nodeList.item(i).getAttributes
        if (attributes != null) {
          0 until attributes.getLength foreach (i => {
            val kv = attributes.item(i).asInstanceOf[Attr]
            if (kv.getName == "selected" && kv.getValue() == "true") {
              b = true
            }
          })
        }
      }
    })
    b
  }

  def getNodeListFromXML(pageDom: Document, xpath: String): AnyRef = {
    val xPath: XPath = XPathFactory.newInstance().newXPath()
    val compexp = xPath.compile(xpath)
    var r: AnyRef = None
    //val node=compexp.evaluate(pageDom)
    if (xpath.matches("string(.*)") || xpath.matches(".*/@[^/]*")) {
      r = compexp.evaluate(pageDom, XPathConstants.STRING)
    } else {
      r = compexp.evaluate(pageDom, XPathConstants.NODESET)
    }
    r
  }


  def getNodeListByXPath(xpath: String, pageDomString: String): List[Map[String, Any]] = {
    val pageDom = toDocument(pageDomString)
    getNodeListByXPath(xpath, pageDom)
  }

  def getNodeListByXPath(xpath: String, pageDom: Document): List[Map[String, Any]] = {
    val node = getNodeListFromXML(pageDom, xpath)
    val nodeMapList = ListBuffer[Map[String, Any]]()
    node match {
      case nodeList: NodeList => {
        log.trace("nodeList length {} with {}", nodeList.getLength, xpath)
        0 until nodeList.getLength foreach (i => {
          val nodeMap = mutable.Map[String, Any]()
          //初始化必须的字段
          nodeMap("name") = ""
          nodeMap("value") = ""
          nodeMap("label") = ""
          nodeMap("x") = "0"
          nodeMap("y") = "0"
          nodeMap("width") = "0"
          nodeMap("height") = "0"


          val node = nodeList.item(i)
          //如果node为.可能会异常. 不过目前不会
          nodeMap("name()") = node.getNodeName
          nodeMap("innerText") = node.getTextContent.trim

          //todo: html有label属性会导致label被覆盖
          //支持导出单个字段
          //nodeMap(node.getNodeName) = node.getNodeValue

          //获得所有节点属性
          val nodeAttributes = node.getAttributes
          if (nodeAttributes != null) {
            0 until nodeAttributes.getLength foreach (a => {
              val attr = nodeAttributes.item(a).asInstanceOf[Attr]
              nodeMap(attr.getName) = attr.getValue
            })
          }

          //todo: 支持selendroid
          //如果是android 转化为和iOS相同的结构
          //name=resource-id label=content-desc value=text
          if (nodeMap.contains("resource-id")) {
            nodeMap("name") = nodeMap("resource-id").toString
          }

          if (nodeMap.contains("content-desc")) {
            nodeMap("label") = nodeMap("content-desc")
          }

          if (nodeMap.contains("text")) {
            nodeMap("value") = nodeMap("text")
          }

          if (nodeMap.contains("latest")) {
            nodeMap("latest") = nodeMap("latest")
          }

          //selenium
          if (nodeMap.contains("href")) {
            if (nodeMap.contains("id")) {
              nodeMap("name") = nodeMap("id")
            }
            if (nodeMap.contains("name")) {
              nodeMap("label") = nodeMap("name")
            }
            nodeMap("value") = nodeMap("innerText")
          }


          val attributesList = getAttributesFromNode(node)
          //必须在xpath前面
          nodeMap("depth") = attributesList.size

          //todo: 改进算法
          //nodeMap("menu") = isMenuFromBrotherNode(node)

          //不同id的祖先可以允许分别遍历，不计算到相同的tagLimit里
          nodeMap("ancestor") = getXPathFromAttributes(attributesList, tagList)

          nodeMap("xpath") = getXPathFromAttributes(attributesList, xpathExpr)


          //为了加速android定位
          if (nodeMap.contains("instance")) {
            nodeMap("instance") = nodeMap("instance")
          }
          //默认true
          nodeMap("valid") = nodeMap.getOrElse("visible", "true") == "true" &&
            nodeMap.getOrElse("enabled", "true") == "true" &&
            nodeMap.getOrElse("valid", "true") == "true"

          //android
          if (nodeMap.contains("bounds")) {
            val rect = nodeMap("bounds").toString.split("[^0-9]+").takeRight(4)
            nodeMap("x") = rect(0).toInt
            nodeMap("y") = rect(1).toInt
            nodeMap("width") = rect(2).toInt - rect(0).toInt
            nodeMap("height") = rect(3).toInt - rect(1).toInt
          }

          if (nodeMap("xpath").toString.nonEmpty) {
            nodeMapList += (nodeMap.toMap)
          }
        })
      }
      case attr: String => {
        //如果是提取xpath的属性值, 就返回一个简单的结构
        nodeMapList += Map("attribute" -> attr)
      }
    }
    log.trace("filted node list length {}", nodeMapList.length)
    nodeMapList.toList
  }

  def getNodeListByKey(key: String, currentPageSource: String): List[Map[String, Any]] = {
    val currentPageDom = toDocument(currentPageSource)
    getNodeListByKey(key, currentPageDom)
  }

  def getNodeListByKey(key: String, currentPageDom: Document): List[Map[String, Any]] = {
    key match {
      //xpath
      case xpath if Array("/.*", "\\(.*", "string\\(/.*\\)").exists(xpath.matches) => {
        getNodeListByXPath(xpath, currentPageDom)
      }
      case regex if regex.contains(".*") || regex.startsWith("^") => {
        getNodeListByXPath("//*", currentPageDom).filter(m => {
          m("name").toString.matches(regex) ||
            m("label").toString.matches(regex) ||
            m("value").toString.matches(regex)
        })
      }
      case str: String => {
        getNodeListByXPath("//*", currentPageDom).filter(m => {
          m("name").toString.contains(str) ||
            m("label").toString.contains(str) ||
            m("value").toString.contains(str)
        })
      }
    }
  }

}
